Step3: 자동차 경주 리뷰 요청드립니다 :)

kwanguuuu opened the pull request:
안녕하세요. 리뷰어님.
step3 자동차 경주 리뷰 요청 드립니다.

막히는 부분도 있는데.. 리뷰 받고 진행하는게 좋을것 같아서요. 🙏🏻
리뷰 부탁드리겠습니다.
 감사합니다!

kwanguuuu committed:
[step3] docs: 기능 요구사항 기반으로 readme 추가 작성

kwanguuuu committed:
[step3] docs-fix: step3 기능 구현 목록 정리.(readme 추가), readme 린트 수정

kwanguuuu committed:
[step3] feat, test:  NumberParser 생성, input 단위 테스트 추가. docs 수정

kwanguuuu committed:
[step3] feat, test:  0-9사이 랜덤수 생성 테스트 작성, 클래스 생성

kwanguuuu committed:
[step3] feat, test: car 객체 생성, 4이상 숫자이면 이동가능 boolean 값 리턴 함수 추가

kwanguuuu committed:
[step3] test: car 객체에 이동/멈춤 기능을 구현함.

- Position.kt 추가
- Car에 이동관련 함수 추가

kwanguuuu committed:
[step3] test: Driver.kt 생성, drive 호출, drive 가능 기준판단 함수를 위임함

kwanguuuu committed:
[step3] test: Driver.kt 호출 테스트 변경

kwanguuuu committed:
[step3] fix, feature: 코드 일부 번경 ( 여러 변경이 있어 아래 기술..)

1. ResultView 생성
2. 패키지 분리 Car 관련, view 관련
3. Driver getMoveResult 응답 변경
4. 테스트 작성

kwanguuuu committed:
[step3] feature: RacingGame.kt 추가. main 추가
fix: ResultView linebreak 목적의 println()삭제

wisemuji commented on src/main/kotlin/step3/racingcar/car/RacingGame.kt:
```diff
@@ -0,0 +1,41 @@
+
+) {
+
+    // TODO: 2022/05/09 RacingGame에서 inputView의 역할을 같이 하고 있는것 같은.. 고민이 있는데 input을 어느시점에 분리하는게 좋을지 고민...
```
<img width="582" alt="image" src="https://user-images.githubusercontent.com/32327475/167540995-5edd370a-68ae-40a9-94f5-a8f29fbbe7db.png">
프로그래밍 요구사항에 맞게 InputView 클래스를 만들어 분리해보면 어떨까요? 이 단계에서 바로 해보실 수 있을 것 같습니다! 고민이 있으시다면 DM으로 이야기 더 나눠볼까요?

wisemuji commented on src/main/kotlin/step3/racingcar/car/Car.kt:
```diff
@@ -0,0 +1,17 @@
+package step3.racingcar.car
+
+class Car {
+    private val drivingRecord: MutableList<Position> = mutableListOf()
+
+    fun getDrivingRecord(): List<Position> {
+        return drivingRecord.toList()
+    }
```
mutable한 List는 내부에 숨기고 외부에는 immutable list만 노출시키신 방향 좋습니다 👍 

backing property를 사용하는 방향으로도 구현할 수 있는데요, 이렇게 구현하면 함수 호출이 아닌 프로퍼티 호출로 record에 접근할 수 있어요.

```kotlin
private val _drivingRecord: MutableList<Position> = mutableListOf()
val drivingRecord get() = _drivingRecord.toList()
```

kwanguuuu replied:
backing property 사용 적용해봤습니다. 프로퍼티로 직접 접근할 필요 있으면 이런식으로 구현이 유용한것 같아요! ㅎ 배워 갑니다...! 🙏🏻

wisemuji commented on src/main/kotlin/step3/racingcar/car/Driver.kt:
```diff
@@ -0,0 +1,27 @@
+        }
+    }
+
+    fun getMoveResult(): Int {
+        return car.getDrivingRecord()
+            .count { it == Position.MOVE }
+    }
```
car의 데이터를 직접 조작하기보단 객체에 메시지를 넘겨보면 어떨까요?

kwanguuuu replied:
Car 객체에 메세지를 넘기면, 인자로 전달한 position enum 값에 일치하는 count를  반환하는 식으로  작성하는것 같은데요

작성해보면..
```kotlin 
// Car.kt

fun recordCount(position: Position) {
    return _drivingRecord.count{ it == position}
}
```

근데 이게 car 가 제공해야할 메서드일까..!? 고민이 되서요

설계시 생각엔 Car는 이동 관련과/ 그 이동 상태만 가진다고 생각했고,
조작은 Driver 클래스에서 하도록 고민하고 시작했었습니당. 

그래서 Car 객체에 직접 메세지를 전달해서 상태를 찾는것 보단 Driver가 Car의 결과로 판단하는게 목적이 일치하지 않나..!? 란 고민이 들었어요!

리뷰어님 의견도 부탁드리겠습니다 ㅎㅎ

wisemuji replied:
객체의 역할과 책임을 분리하는게 쉽지 않으실텐데요, 
> 설계시 생각엔 Car는 이동 관련과/ 그 이동 상태만 가진다고 생각했고,
조작은 Driver 클래스에서 하도록 고민하고 시작했었습니당.

**Car는 이동 관련과/ 그 이동 상태만 가진다고** 말씀해주셨는데 getMoveResult()의 역할이 **차가 지금까지 움직인 거리**를 리턴하는 것이라고 생각해보면 Car가 그 역할을 가지는게 어색하지 않다고 생각합니다. 해당 함수가 Car를 조작하는 역할이 아닌, **차에게 지금까지 움직인 거리를 리턴해!** 라고 메시지를 전달하는 역할을 한다고 생각해보면 어떨까요? 🙂

kwanguuuu replied:
오 넵 주신 의견 이해했습니다. ㅎㅎ 그리고 반영해보려고 하는데요!
<img width="707" alt="스크린샷 2022-05-11 오후 11 45 59" src="https://user-images.githubusercontent.com/18343294/167878295-de1b038f-2394-4a2d-93fb-140c64a2f14f.png">

Car 객체에 currentDrivingRecord라는 프로퍼티를 갖고 이 걸 사용하면 리뷰어님의 의도대로 표현할 수 있을거 같아요!
하나 궁금한건, 하나의 backing property를 이용해 여러 프로퍼티를 사용하는건 어떻게 생각하세요?

위의 값을 굳이 함수로 표현하지 않아도 될것 같아서 backing property 를 이용한 프로퍼티 추가를 선택했어요.!

근데 backing property 변경이 있다면, 두 퍼블릭 프로퍼티도 변경이 함께 일어나니까 이런 코드가 실제로 많은지& 좋은..방향이라 할 수 있는지 궁금했습니다.ㅎ.ㅎ

wisemuji replied:
작성해주신 성격의 코드는 실제로도 많이 사용합니다! 오히려 해당 성격의 값을 프로퍼티로 설정하신건 좋은 선택이라고 생각해요!

아래의 글도 도움이 되면 좋겠어요 👍 
https://blog.kotlin-academy.com/kotlin-should-i-define-function-or-property-6786951da909

wisemuji replied:
> 근데 backing property 변경이 있다면, 두 퍼블릭 프로퍼티도 변경이 함께 일어나니까 이런 코드가 실제로 많은지& 좋은..방향이라 할 수 있는지 궁금했습니다.ㅎ.ㅎ

나중에 Coroutine과 Flow를 사용하시게 될 기회가 생기신다면 비슷한 성격의 코드가 익숙해지실거예요 ㅎㅎ

wisemuji commented on src/main/kotlin/step3/racingcar/car/RacingGame.kt:
```diff
@@ -0,0 +1,41 @@
+            println("시도할 횟수는 몇 회 인가요?")
+            val trialCount = InputParser.toInt(readln())
+
+            val drivers = mutableListOf<Driver>()
+            repeat(numberOfCar) {
+                drivers.add(Driver(Car()))
+            }
+
```
https://kotlinlang.org/docs/constructing-collections.html#initializer-functions-for-lists

kotlin의 List constructor-like function을 사용하면 mutableList 생성 없이 한번에 원하는 리스트를 생성할 수 있습니다.

kwanguuuu replied:
유용한 방식 같아요..! 실제로 mutableList 로 만든뒤 조작하는 코드를 많이 작성하고 있었는데, 적극 적용해보겠습니다. 감사합니다

wisemuji commented on src/main/kotlin/step3/racingcar/view/ResultView.kt:
```diff
@@ -0,0 +1,11 @@
+        repeat(result) {
+            print(DASH_VIEW)
+        }
+    }
```
함수명이 내부의 동작을 나타내기에 모호한 것 같습니다. 어떤 함수명이 같은 동료 개발자도 이해할 수 있는 좋은 함수명일까요?

kwanguuuu replied:
`showRecordsAsDash` 로 변경했는데요, 기록을 대시뷰로 나타난대는 이름으로 수정했습니다.!

해당 함수의 인자가 Int -> List<Int>로 변경됐는데, ( 경주 결과들을 한번에 받아 출력하냐고 )
이 부분 변경이 view 구현 방식이 바뀐건가!? 란 생각은 드네요

wisemuji commented on src/test/kotlin/step3/racingcar/NumberInputTest.kt:
```diff
@@ -0,0 +1,20 @@
+
+class NumberInputTest {
+    @Test
+    fun `횟수 n을 입력 받는다 n은 정수 int type`() {
```
테스트명을 작성할 때 어떤 것을 검증하려고 하는지 더 잘 드러날 수 있도록 작성해보면 어떨까요? (이하 동일) 예를 들어 다음과 같이 작성할 수 있을 것 같아요.
```
입력받은 시도 횟수가 정수가 아니라면 IllegalArgumentException 발생
```

kwanguuuu replied:
이름을 너무 대충 지었네요....ㅎㅎ; `숫자인 문자열을 받으면 int로 변환한다` 로ㅓ 수정했습니다!

https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r869066704
이 리뷰 코멘트도 함꼐 적용했어요

성공/실패 케이스를 분리하니 실패케이스는 @ParameterizedTest로 작성했습니다.
+검증방식이 다를땐 @ParameterizedTest를 적용하기 힘든것은 맞나요..!?

wisemuji replied:
네! 질문주신 것처럼 검증방식이 다를땐 활용하기 어렵습니다 😭  함수 적절히 잘 분리해주셨습니다 👍

wisemuji commented on src/test/kotlin/step3/racingcar/NumberInputTest.kt:
```diff
@@ -0,0 +1,20 @@
+        val nullInput = null
+        val blinkInput = ""
+
+        assertThat(InputParser.toInt(intInput)).isEqualTo(4)
+        assertThatThrownBy { InputParser.toInt(stringInput) }.isInstanceOf(IllegalArgumentException::class.java)
+        assertThatThrownBy { InputParser.toInt(nullInput) }.isInstanceOf(IllegalArgumentException::class.java)
+        assertThatThrownBy { InputParser.toInt(blinkInput) }.isInstanceOf(IllegalArgumentException::class.java)
```
매번 이렇게 각 케이스별로 assertThatThrownBy를 호출해야 할까요? [ParameterizedTest](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests)를 활용하시면 더욱 깔끔하게 구현할 수 있을 것 같습니다.

kwanguuuu replied:
`Question`:  테스트에 assertThat, assertThrownBy가 섞여 있음에도 parameterizedTest를 활용할 수 있나요?? 검증 방식이 달라서 각각 작성했는데, 요령이 있으면 전수 부탁드립니당...!

wisemuji replied:
성공한 경우의 테스트와 실패한 경우의 테스트를 분리해보시면 어떨까요!

wisemuji commented on src/test/kotlin/step3/racingcar/car/CarTest.kt:
```diff
@@ -0,0 +1,37 @@
+        // then
+        assertThat(recored2.size).isEqualTo(1)
+        assertThat(recored2[0]).isEqualTo(Position.STOP)
+    }
```
두 개의 케이스에 대한 테스트가 하나의 테스트 함수에서 검증되고 있는데요, 예를 들어 stop() 함수에 문제가 생겼다면 move()가 정상적으로 동작함에도 이 테스트는 같이 실패해버려요. 의미있는 단위로 테스트를 나눠서 진행해보면 어떨까요?

kwanguuuu replied:
오 넵 서로 로직이 영향이 있을 수 있는것 같습니다. stop, start 로직 분리하고 결과 상태를 확인하는 테스트를 추가했고, 위의 테스트는 삭제했습니다!

wisemuji commented on src/test/kotlin/step3/racingcar/car/CarTest.kt:
```diff
@@ -0,0 +1,37 @@
+    @Test
+    fun `여러번 차 이동멈춤을 시도한다`() {
+        val car = Car()
+        // 이동멈춤 권한을 driver에게 이동
```
해당 테스트 함수는 제거하셔도 좋을 것 같습니다! outdated 된 로직에 대한 테스트 함수는 유지보수되지 않을 가능성이 높습니다.

wisemuji commented on src/test/kotlin/step3/racingcar/view/ViewTest.kt:
```diff
@@ -0,0 +1,11 @@
+    fun `ResultView 객체를 생성, 입력값 만큼 -를 print한다`() {
+        ResultView.show(5)
+    }
+}
```
<img width="582" alt="image" src="https://user-images.githubusercontent.com/32327475/167540995-5edd370a-68ae-40a9-94f5-a8f29fbbe7db.png">
프로그래밍 요구사항에 맞게 UI 로직은 단위 테스트에서 제외하고 핵심 도메인 로직을 테스트하는데에 집중해보면 어떨까요?

wisemuji commented on src/test/kotlin/step3/racingcar/car/DriverTest.kt:
```diff
@@ -0,0 +1,25 @@
+
+        driver.drive(boundaryNumber)
+        assertThat(driver.getMoveResult()).isEqualTo(2)
+    }
```
랜덤이 포함된 요구사항에 대한 테스트 로직을 잘 작성해주셨습니다 👍 

다만 하나 고민해보시면 좋을 점이 있어요. 현재 Driver 객체에게 차를 움직이라는 메시지를 전하기 위해서는 내부 구현이 **무작위 값이 4 이상일 경우 움직인다** 라는 사실을 이해한 상태여야 하는데요, 
```kotlin
driver.drive(4)  // 움직임
driver.drive(3)  // 안움직임
```
이러한 성격의 함수는 다음과 같은 단점이 존재합니다.
1. 유지보수하기 까다롭습니다. - 만약 이 요구사항이 변경되는 경우에는 호출부에서 전부 바꿔줘야 합니다. 
1. 내부 로직을 이해하지 않으면 (컨텍스트를 모르는 동료 개발자들은) 해당 함수를 사용할 수 없습니다.

객체에게 '**움직여**'라는 메시지만 전할 수 있도록 구현해보세요!

kwanguuuu replied:
게더 감사합니다. 뭔가 방향을 못잡고 있었 던 것 같은데 많은 도움 됐어요!
- 전략추가 https://github.com/next-step/kotlin-racingcar/pull/482/commits/60467c2364d6ab2b99133e1fca223845a5961f4f
- 코드 적용&테스트 https://github.com/next-step/kotlin-racingcar/pull/482/commits/71b5dcd9f6df619a3fdbdbb02a49140a868ee3a3

MoveStrategy란 인터페이스에 isMovable() 함수를 두고, 구현체에서 그 함수가 어떤 조건으로 movable 가능한지를 나타내도록 변경했습니다.
생성한 전략을 도메인(Driver)가 주입받게 만드니, 전략이 움직일 수 있는지 여부만 확인하고, drive 호출이 바로 되네요..!

이렇게 되면 다른 전략이 적용 되도, isMovable이 true가 되는 조건을 전략에서도 확인할 수 있고, driver 객체도 수정을 하지 않아도 되서 좋은것 같네요..!

p.s 이 내용이 맞죠!?

wisemuji replied:
전략 패턴에 대해 이해하고 적절히 잘 적용해주셨어요 👍  

> 이렇게 되면 다른 전략이 적용 되도, isMovable이 true가 되는 조건을 전략에서도 확인할 수 있고, driver 객체도 수정을 하지 않아도 되서 좋은것 같네요..!

네 맞습니다!! 로직 실행은 인터페이스에 의존을 시키고 인터페이스를 구현한 로직들을 전달해줌으로써 테스트하기 쉬운, 유연한 코드 구조를 가질 수 있습니다. 테스트 가능한 코드를 만들기 위한 방법은 앞으로의 강의에서도 계속 학습하실 수 있을테니 또 활용해볼 기회가 있으시면 좋겠어요!

wisemuji reviewed (changes requested):
Step3 구현을 잘 해주셨습니다 👏 

객체를 의미있는 단위로 나누고 각각의 객체에 역할과 책임을 적절히 부여한 구조가 인상깊었습니다 💯 
고민해보시면 좋을만한 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

kwanguuuu committed:
[step3] fix: 미사용 테스트 코드 삭제

- pr반영: https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r869069608

kwanguuuu committed:
[step3] refactor: get 메서드 backing property 사용하도록 변경

- pr반영: https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r869023804

kwanguuuu committed:
[step3] refactor,test: construct-like function 학습 테스트, 리스트 생성방식 변경 적
용

- pr 리뷰 반영: https://github.com/next-step/kotlin-racingcar/pull/482#discussion_r869034924

kwanguuuu committed:
[step3] refactor: InputView 생성, RacingGame에 존재하던 Input 관련 로직 삭제

kwanguuuu committed:
[step3] refactor, test: 차 이동&멈춤 테스트 케이스 각각 분리, 기존 테스트 삭제.
Car 객체에 현재 위치를 나타내는 프로퍼티 추가(currentDrivingRecord)

kwanguuuu committed:
[step3] test, refactor: Driver가 차의 현재 위치 값을 받는 테스트 작성, driver가 car의 현재 위치 받도록 수정

kwanguuuu committed:
[step3] refactor: RacingGame이 resultView 관련 코드 사용하지 않도록 삭제

kwanguuuu committed:
[step3] refactor: ResultView 메서드명 변경, 출력 로직 수정

kwanguuuu committed:
[step3] feat: racingGame이 실행가능 횟수를 판단하는 함수 추가

kwanguuuu committed:
[step3] feat: racingGame 게임 진행 횟수 감소 추가.

kwanguuuu committed:
[step3] refactor: ResultView return type 수정, 관련 함수명 수정

kwanguuuu committed:
[step3] feat: MoveStrategy 관련 객체 추가

kwanguuuu committed:
[step3] refactor

1. moveStrategy에 따른 drive 가능 여부 판단 테스트 추가.
2. 테스트용 더미 MoveStrategy 구현체 추가
3. moveStrategy 추가로 발생한 생성자, 테스트 코드 컴파일 에러 수정
4. 이전에 놓친 ViewTest 코드 컴파일 에러 수정

kwanguuuu committed:
[step3] feat: prod 코드에 있던 AlwaysMoveStrategy.kt 삭제

kwanguuuu committed:
[step3] fix: NumberInputTest 테스트  성공/실패 케이스 분리

kwanguuuu commented:
@wisemuji 
리뷰 감사합니다! 코멘트 주신 리뷰 수정 방향 & 의견들 각각 모두 추가로 달아놨어요.~
이번 request에도 이상한 부분 있으면 코멘트 부탁드리겠습니다 😇

kwanguuuu committed:
[step3] refactor: 미사용 함수 삭제 - Car.stop() -> 굳이 스톱을 호출할 이유가 없는거 같아요..

wisemuji commented on src/main/kotlin/step3/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,15 @@
+import step3.racingcar.InputParser
+
+object InputView {
+    fun showNumberOfCarInitView(): Int {
+        println("자동차 대수는 몇 대 인가요?")
+        return InputParser.toInt(readln())
+    }
+
+    fun showTrialCountInitView(): Int {
+        println("시도할 횟수는 몇 회 인가요?")
+        return InputParser.toInt(readln())
+    }
```
`show~`의 형태로 명명하셨지만 실제로는 input 받고 값을 반환하는 역할을 하고 있어요. 함수명만 보고도 동작을 예상할 수 있도록 명명해보세요!

wisemuji commented on README.md:
```diff
@@ -1,12 +1,41 @@
+
+- [x] 문자열 계산은 사칙 연산 순서가 아닌, 입력 순서대로 작동하도록 하기
+
+## step3 기능 요구사항 정리
+
+- 주어진 횟수 동안 n대의 자동차는 전진 또는 멈출 수 있다.
+    - [ ] 횟수 n을 입력 받는다
+    - [ ] 자동차는 전진or멈춤 한다.
+- 사용자는 몇대의 자동차로 몇번의 이동을 할 것인지 구할 수 있다.
+    - [ ] 자동차 n개를 입력 받는다.
+    - [ ] 
+- 전진하는 조건은 0에서 9 사이 무작위 값을 구한 후 무작위 값이 4 이상일 경우다.
+    - [ ] 0-9 사이 무작위 값을 구한다
+    - [ ] 무작위 값이 4 이상이다
+    - [ ] 4 이상이면 전진한다.
+- 자동차의 상태를 화면에 출력한다. 어느 시점에 출력할 것인지 제약은 없다.
+    - [ ] 진행 상태를 화면에 출력한다..
```
구현이 완료된 요구사항은 체크표시 해주시면 좋을 것 같아요!

wisemuji commented on src/main/kotlin/step3/racingcar/InputParser.kt:
```diff
@@ -0,0 +1,8 @@
+package step3.racingcar
```
이제 4단계를 진행하시게 되면 패키지 명이 바뀔 것 같은데요, racingcar로만 명명하셔도 괜찮다고 생각합니다!

wisemuji commented on src/main/kotlin/step3/racingcar/car/Car.kt:
```diff
@@ -0,0 +1,10 @@
+
+class Car {
+    private val _drivingRecord: MutableList<Position> = mutableListOf()
+    val currentDrivingRecord get() = _drivingRecord.count { it == Position.MOVE }
```
drivingRecord 리스트를 드러내지 않고 필요한 값만 노출하는 방향 좋습니다 👍

wisemuji commented on src/main/kotlin/step3/racingcar/car/RacingGame.kt:
```diff
@@ -0,0 +1,44 @@
+    companion object {
+        private const val GAME_END_COUNT = 0
+
+        fun init(numberOfCar: Int, trialCount: Int, strategy: MoveStrategy): RacingGame {
+            val drivers = List(numberOfCar) { Driver(Car(), strategy) }
+            return RacingGame(drivers, trialCount)
+        }
```
init function 함수를 따로 만들기보다 [secondary constructor](https://kotlinlang.org/docs/classes.html#secondary-constructors)를 활용해보시면 어떨까요?

kwanguuuu replied:
설계 할 땐 init()이란 정적 함수로 구현하는게 쿨해 보였는데...ㅋㅋ constructor 사용해도 될것 같습니다 ~ 그렇게 구현할게요

kwanguuuu replied:
secondary construct를 사용해서 구현해 봤는데요,
```kotlin
construct RacingGame(numberOfCar: Int, trialCount: Int, strategy: MoveStrategy): this( List(numberOfCar) { Driver(Car(), strategy) }, trialCount )
```
단순 이동시키니..
란 엄청 긴 생성자 이름이 나왔네용..
이게 더 맘에 안들어서.. 주생성자로 변경했습니닷!

wisemuji commented on src/main/kotlin/step3/racingcar/car/RacingGame.kt:
```diff
@@ -0,0 +1,44 @@
+        return drivers.map { it.getMoveResult() }
+    }
+
+    fun isPlayable(): Boolean {
+        if (trialCount <= GAME_END_COUNT) {
+            throw GameOverException()
+        }
+        return true
+    }
```
isPlayable() 함수 시그니처만 봤을 때는 플레이 가능한지 여부를 Boolean 값으로 리턴해줄 것 같은데 
```kotlin
fun isPlayable(): Boolean
```
실제로는 플레이가 불가능하면 Exception을 던지고 있어요. 함수 시그니처와 내부 동작을 통일하면 어떨까요?

wisemuji commented on src/main/kotlin/step3/racingcar/car/strategy/MoveStrategy.kt:
```diff
@@ -0,0 +1,5 @@
+package step3.racingcar.car.strategy
+
+interface MoveStrategy {
+    fun isMovable(): Boolean
+}
```
참고로 이 SAM(== Single Abstract Method, An interface with only one abstract method) 형식의 interface는 다음과 같이 [functional interface](https://kotlinlang.org/docs/fun-interfaces.html)로도 만들 수 있습니다!
[as-is]
```kotlin
interface MoveStrategy {
    fun isMovable(): Boolean
}

// usage
Driver(
    Car(),
    object : MoveStrategy {
        override fun isMovable(): Boolean {
            return false
        }
    }
)
```
[to-be]
```kotlin
fun interface MoveStrategy {
    fun isMovable(): Boolean
}

// usage
Driver(Car()) { false }
```

호출부가 훨씬 간결해졌죠? 꼭 반영하실 필요는 없습니다!

wisemuji commented on src/test/kotlin/step3/racingcar/NumberInputTest.kt:
```diff
@@ -0,0 +1,25 @@
+    @NullSource
+    @ValueSource(strings = ["문자", " ", "\t"])
+    fun `숫자가 아닌 문자열을 입력 받으면 에러를 발생시킨다`(value: String?) {
+        assertThatThrownBy { InputParser.toInt(value) }.isInstanceOf(IllegalArgumentException::class.java)
```
```suggestion
        assertThrows<IllegalArgumentException> { InputParser.toInt(value) }
```
코틀린 전용 assertThrows 함수도 있으니 활용해보시면 좋을 것 같아요!

wisemuji commented on src/test/kotlin/step3/racingcar/car/CarTest.kt:
```diff
@@ -0,0 +1,18 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class CarTest {
+    @Test
+    fun `차를 이동 후 현재 위치결과를 반환한다`() {
+        // given
+        val testCar = Car()
+
+        // when
+        testCar.move()
+
+        // then
+        assertThat(testCar.currentDrivingRecord).isEqualTo(1)
+    }
+}
```
작성하시는 리뷰 스타일을 보니 Kotest의 [BehaviorSpec](https://kotest.io/docs/framework/testing-styles.html#behavior-spec)이 이런 given-when-then의 상황에서 요긴하게 쓰일 수 있어요!
이것도 꼭 반영하실 필요는 없고 참고용입니다! 😆

wisemuji commented on src/test/kotlin/step3/racingcar/car/DriverTest.kt:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Test
+    fun `드라이버의 전략이 false면 차가 움직일 수 있다`() {
```
```suggestion
    fun `드라이버의 전략이 false면 차가 움직일 수 없다`() {
```
작은 오타가 있네요!

wisemuji commented on src/test/kotlin/step3/racingcar/car/NumberGenerateTest.kt:
```diff
@@ -0,0 +1,17 @@
+
+class NumberGenerateTest {
+
+    @RepeatedTest(100)
+    fun `0-9 사이 임의의 정수를 생성한다`() {
+        val testRange = 1..9
+        val testList = listOf(1, 2, 3, 4, 5, 6, 7, 8, 9)
+
+        assertThat(RandomNumberGenerator.generate()).isIn(testRange)
+        assertThat(RandomNumberGenerator.generate()).isIn(testList)
+    }
```
랜덤한 로직 자체를 테스트하는건 **단위 테스트**의 성격에 맞지 않을 수 있습니다.

이 글이 도움이 되면 좋겠습니다:
https://softwareengineering.stackexchange.com/questions/356456/testing-a-function-that-uses-random-number-generator

> The important thing to note is that you are not proving randomness. You are ensuring that your method behaves as expected.

> If you were testing randomness, you would need to do some statistical analysis to prove the spread of random numbers was as expected, etc. Those kinds of tests would require "proof" in the mathematical sense of the word. Unit tests aren't the right tool for that job.

wisemuji commented on src/test/kotlin/step3/racingcar/view/ViewTest.kt:
```diff
@@ -0,0 +1,11 @@
+
+import org.junit.jupiter.api.Test
+
+class ViewTest {
+
+    @Test
+    fun `ResultView 객체를 생성, 입력값 만큼 -를 print한다`() {
+        ResultView.showRecordsAsDash(listOf(5))
+    }
+}
```
UI에 대한 테스트는 작성하지 않으셔도 됩니다!
<img width="633" alt="image" src="https://user-images.githubusercontent.com/32327475/168090374-d891b887-0cda-4cfb-b342-d50df3348041.png">

wisemuji reviewed (approved):
코멘트를 꼼꼼하게 잘 반영해주셨습니다 💯 

객체에 메시지를 던지는게 익숙해지신게 보입니다!! 함수가 하나의 역할만 하도록 구현해주신 것도 너무 좋았습니다 👍 

조금 고민해보시면 좋을 코멘트를 남겼으니 Step4에서 같이 반영해주시면 되겠습니다.
저번처럼 이해가 되지 않는게 있으시다면 편하게 DM 주세요 ㅎㅎ
빠른 미션 진행을 위해 이 PR은 머지하겠습니다 🙇
